var/spack/repos/builtin/packages/trilinos/package.py

```diff
@@ -678,20 +680,41 @@ def cmake_args(self):
                     '-DTrilinos_ENABLE_Fortran=ON'
                 ])
 
-        # Explicit instantiation
-        if '+instantiate' in spec:
+        # Instantiation in Tpetra and Teuchos
+        # As of 12.12.1, DTrilinos_ENABLE_EXPLICIT_INSTANTIATION does not really build with
+        # extra instantiations of Tpetra, see
+        # https://github.com/trilinos/Trilinos/issues/3566 and
+        # https://github.com/trilinos/Trilinos/issues/3566
+        # To avoid build issues but still support additional types, separate
+        # explicit instantiation setting and Tpetra's _INST_
+        if '+explicit_instantiation' in spec:
             options.extend([
                 '-DTrilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON'
             ])
+
+        if '+instantiate' in spec:
             if '+tpetra' in spec:
                 options.extend([
                     '-DTpetra_INST_DOUBLE:BOOL=ON',
-                    '-DTpetra_INST_INT_LONG:BOOL=ON'
+                    '-DTpetra_INST_INT_UNSIGNED_LONG=ON',
+                    '-DTpetra_INST_INT_LONG:BOOL=ON',
                     '-DTpetra_INST_COMPLEX_DOUBLE=%s' % (
                         'ON' if '+instantiate_cmplx' in spec else 'OFF'
-                    )
+                    ),
+                    '-DTpetra_INST_COMPLEX_FLOAT=%s' % (
+                        'ON' if '+instantiate_cmplx' in spec else 'OFF'
+                    ),
+                    '-DTpetra_INST_FLOAT=ON'
                 ])
 
+                if '+teuchos' in spec:
+                    options.extend([
+                        '-DTeuchos_ENABLE_COMPLEX=%s' % (
```

Comment:
If I understand it right, `Teuchos_ENABLE_COMPLEX` has nothing to do with instantiations but rather enabling complex numbers. You can still use complex even without explicit instantiations.